### ISSUE TITLE
Namespace macros

### ISSUE THREAD

**Issue by viboes** [boostorg/hana#288]
What about changing

```
BOOST_HANA_NAMESPACE_BEGIN
```

by

```
namesapce  BOOST_HANA_NAMESPACE_OPEN {
```

and

```
BOOST_HANA_NAMESPACE_END
```

by

```
} BOOST_HANA_NAMESPACE_CLOSE
```

This reduces the syntactical noise of the macros, the braces are pair wised enabling folding in some editors.

``` c++
namespace  BOOST_HANA_NAMESPACE_OPEN {
} BOOST_HANA_NAMESPACE_CLOSE
```

**Comment by ldionne** [comment]
I don't think this reduces the syntactical noise of the macros (quite the contrary, in fact), because you have more symbols to type and you have to think about how the macro interacts with these symbols. The ease of interaction with editors is a valid point. What editor are you using?

**Comment by ldionne** [comment]
But regardless, why would you want to fold a whole namespace, since it basically means folding the whole contents of the file given the structure of Hana's headers?

**Comment by viboes** [comment]
I'm using Eclipse. You are right that most of the headers (if not all) wouldn't profit from unfolding :( 

Be free to close it if you don't find an advantage.

**Comment by ldionne** [comment]
All I'm trying to see if whether that provides an advantage to users of some editor(s). For example, I know Sublime Text -- which I use -- wouldn't benefit from it much, but if this simple change is a big plus for many potential users, then it might be worth it to change. Is there an advantage to using what you propose in Eclipse?

**Comment by ricejasonf** [comment]
I think I remember that the macro was put in place for forwards compatibility with the use of inline namespaces or something. If the macro is causing problems, maybe a simple script or refactoring tool could get the job done when the time comes.

**Comment by ldionne** [comment]
The macro is not causing any problem to me, at least. @viboes Is there any advantage to doing what you propose in Eclipse?

**Comment by ldionne** [comment]
Closing for lack of feedback.

**Comment by Aster89** [comment]
Hello there! @ricejasonf linked this off [my question on stackoverflow](https://stackoverflow.com/questions/65741227/why-does-boost-hana-use-macros-for-opening-closing-namespaces?noredirect=1#comment116240629_65741227), hence I write to give my feedback.

Writing a namespace in the non-macro way, if the editor is able to take advantage of it, allows 3 things:

  - folding, which I really don't care about, honestly;
  - jumping from opening to closing curly braces;
  - being with the cursor on the closing/opening curly brace, the corresponding opening/closing curly brace can be shown in a popup.

The last two points are shown in the following screencast, where I first attempt jumping from macro to macro, which obviously doesn't work (resulting in jumping from the first jump-able thing down the file), and then edit the file to show the jumping and popup.

[![asciicast](https://asciinema.org/a/ZulWI2Ez29GhuvQbh9cZo7FLL.svg)](https://asciinema.org/a/ZulWI2Ez29GhuvQbh9cZo7FLL)

**Comment by ldionne** [comment]
As Jason said elsewhere, this was initially for forward compatibility with inline namespaces. This hasn't been relevant so far, and I don't think it will, so I'll make the change.
